app/services/facemasks.ts

```diff
@@ -0,0 +1,617 @@
+    return new Promise((resolve, reject) => {
+      const asyncReads = missing.map(uuid => this.readFile(uuid));
+      Promise.all(asyncReads).then(results => {
+        resolve();
+      }).catch(err => {
+        this.notifyFailure();
+        reject();
+      });
+    });
+  }
+
+  updateMasks(data: IFacemask[]) {
+    const needsUpdate = data.reduce((result, mask) => {
+      if (this.state.modtimeMap[mask.uuid].modtime < (mask.modtime - 3600)) {
+        result.push({ uuid: mask.uuid, intro: mask.is_intro });
+      }
+      return result;
+    }, []);
+
+    const downloads = needsUpdate.map(mask => this.downloadAndSaveModtime(mask.uuid, mask.intro, true));
+
+    Promise.all(downloads).then(responses => {
+      this.activate();
+    }).catch(err => {
+      console.log(err);
+      this.notifyFailure();
+    });
+  }
+
+  readFile(uuid: string) {
+    const maskPath = this.libraryPath(uuid);
+    return new Promise((resolve, reject) => {
+      fs.readFile(maskPath, 'utf8', (readError, data) => {
+        if (readError) reject(readError);
+        try {
+          const maskData = JSON.parse(data);
+          this.ADD_MODTIME(uuid, maskData.modtime, maskData.is_intro);
+          resolve();
+        } catch (parseError) {
+          fs.unlinkSync(maskPath);
+          reject(parseError);
+        }
+      });
+    });
+  }
+
+  checkDownloaded(uuid: string) {
+    const maskPath = this.libraryPath(uuid);
+    return !fs.existsSync(maskPath);
+  }
+
+  enableMask(uuid:string) {
+    this.downloadMask(uuid).then(modtime => {
+      this.ADD_MODTIME(uuid, modtime, false);
+      this.fetchInstallUpdate(uuid);
+    });
+  }
+
+  // Try to download a mask, resolve whether operation was successful or not
+  downloadAndSaveModtime(uuid: string, intro: boolean, update = false): Promise<any> {
+    return new Promise((resolve, reject) => {
+      this.downloadMask(uuid, update).then(modtime => {
+        if (modtime) {
+          this.ADD_MODTIME(uuid, modtime, intro);
+          this.fetchInstallUpdate(uuid);
+        }
+        resolve();
+      }).catch(err => {
+        reject(err);
+      });
+    });
+  }
+
+  // Returns a promise that resolves with a masks modtime or false if it is already downloaded
+  downloadMask(uuid: string, update = false): Promise<number> {
+    const maskPath = this.libraryPath(uuid);
+
+    // Don't re-download the facemassk if we have already downloaded it unless it needs to be updated
+    if (!update && fs.existsSync(maskPath)) {
+      return Promise.resolve(null);
+    }
+
+    let fileContent = '';
+    return new Promise((resolve, reject) => {
+      this.ensureFacemasksDirectory();
+
+      https.get(this.libraryUrl(uuid), response => {
+        const writeStream = fs.createWriteStream(maskPath);
+        response.on('data', chunk => fileContent += chunk);
+        writeStream.on('finish', () => {
+          try {
+            const data = JSON.parse(fileContent) as IFacemask;
```

Comment:
Cool so here we are downloading the file to disk and also keeping a copy in memory at the same time to pull out the modtime, right?  👍

Reply:
Yup.